python/ray/autoscaler/v2/scheduler.py

```diff
@@ -0,0 +1,346 @@
+
+# ============= Resource Scheduling Service API =======================
+#
+#  ResourceSchedulerService is a service that schedules resource bundles
+#  to nodes. It's used by the autoscaler to schedule resource bundles
+#  to determine the desired cluster size to satisfy the current resource
+#  demands.
+#
+
+
+@dataclass
+class NodeTypeConfig:
+    # Node type name
+    name: NodeType
+    # The minimal number of workers to be launched for this node type.
+    min_workers: int
+    # The maximal number of workers can be launched for this node type.
+    max_workers: int
+    # The total resources on the node.
+    resources: Dict[str, float] = field(default_factory=dict)
+    # The labels on the node.
+    labels: Dict[str, str] = field(default_factory=dict)
+
+
+@dataclass
+class ClusterConfig:
+    # The node type configs.
+    node_type_configs: Dict[NodeType, NodeTypeConfig] = field(default_factory=dict)
+    # The max number of worker nodes to be launched for the entire cluster.
+    max_num_worker_nodes: Optional[int]
+
+
+@dataclass
+class SchedulingRequest:
+    # TODO: This prob could be refactored into the ClusterStatus data class later.
+    # The current ray resource requests.
+    resource_requests: List[ResourceRequestByCount] = field(default_factory=list)
+    # The Gang resource requests.
+    gang_resource_requests: List[GangResourceRequest] = field(default_factory=list)
+    # cluster resource constraints.
+    cluster_resource_constraints: List[ClusterResourceConstraint] = field(
+        default_factory=list
+    )
+    # The ray nodes
+    node_states: List[NodeState] = field(default_factory=list)
+    # The current list of instances.
+    current_instances: List[Instance] = field(default_factory=list)
+    # The resource schedule config.
+    schedule_config: ClusterConfig
+
+
+@dataclass
+class SchedulingReply:
+    # The infeasible resource bundles.
+    infeasible_resource_requests: List[ResourceRequestByCount] = field(
+        default_factory=list
+    )
+    # The infeasible gang resource bundles.
+    infeasible_gang_resource_requests: List[GangResourceRequest] = field(
+        default_factory=list
+    )
+    # The infeasible cluster resource constraints.
+    infeasible_cluster_resource_constraints: List[ClusterResourceConstraint] = field(
+        default_factory=list
+    )
+    # The target cluster shape, given the current resource demands and instances.
+    # Key is the node type name, value is the number of nodes.
+    # This is needed to prevent autoscaler terminating nodes needed for cluster
+    # constraints.
+    # Note this might be "smaller" than the current cluster shape, since there
+    # could be cluster constraints enforced, e.g. a newly updated max_workers value
+    # would result in a target count smaller than the current count of the node type.
+    target_cluster_shape: Dict[NodeType, int]
+    # New launch needed to be launched.
+    to_launch_nodes: Dict[NodeType, int]
```

Comment:
This is not needed since the information can be derived from `target_cluster_shape`